### ISSUE TITLE
allow sharing a coder app or a port with specfic coder users

### ISSUE THREAD

**Issue by matifali** [coder/coder#11201]
## Context

Currently, we have three options for sharing a [`coder_app`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/app).

> [`share`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/app#share) (String) Determines the "level" which the application is shared at. Valid levels are "owner" (default), "authenticated" and "public". Level "owner" disables sharing on the app, so only the workspace owner can access it. Level "authenticated" shares the app with all authenticated users. Level "public" shares it with any user, including unauthenticated users. Permitted application sharing levels can be configured site-wide via a flag on coder server (Enterprise only).

I think the authenticated permission is too broad, and most users would not wish to expose their workspace to everyone in the organization. 

## Suggestion

Allow exposing apps to a single user by specifying the target `username`.

### Proposal 1

This can be achieved by setting `share = "matifali"` in the [`coder_app`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/app) resource.

A template admin can configure it as a [`coder_parameter`](https://registry.terraform.io/providers/coder/coder/latest/docs/data-sources/parameter), and the user can configure it while creating or updating their workspace.

```hcl
data "coder_parameter" "share_code_server" {
  name         = "share_code_server"
  display_name = "Share code-server"
  description  = "Sharing code-server with user: e.g matifali, bpmct"
  type         = "string"
  mutable      = true
  default      = ""
}

resource "coder_app" "code-server" {
  count        = local.code-server-count
  agent_id     = coder_agent.main.id
  display_name = "VS Code Web"
  slug         = "code-server"
  url          = "http://localhost:13337"
  icon         = "/icon/code.svg"
  subdomain    = true
  share        = data.coder_parameter.share_code_server.value == "" ? "owner" : data.coder_parameter.share_code_server.value
}
```
This should then share this coder_app with the username specified for the `share` property. 

---
### Proposal 2

Allow doing this from the workspace page without any template admins, and it could be disabled/enabled deployment-wide using a flag e.g., `CODER_DISBALE_SHARING_CODER_APPS=false/true`
This could be done along with #10908 

cc: @sharkymark and @coder/pms for commnets

**Comment by sharkymark** [comment]
I agree fine-grained sharing is needed, but doing this in Terraform requires a template admin.  Maybe a group.

But I think all of this sharing should optionally be done at the user-level while in the workspace.

**Comment by matifali** [comment]
@Emyrk what do you think about this? Specifically Proposal 2.

**Comment by Emyrk** [comment]
I don't think this should be done in the template. This should be handled by Coder outside of any template terraform file imo and handled in our authorization engine.

There is an effort to redo our rbac that would enable this more easily. If it is pressing, we can bolt it onto our existing system similar to how we did groups.

**Comment by matifali** [comment]
We plan to do it outside terraform with RBAC.

**Comment by Emyrk** [comment]
@matifali I just noticed that Proposal 2 does not include proposal #1 :facepalm:.

Yes, I think prop 2 is the way to go. I would add that sharing should be a toggle feature on the template as well.

**Comment by glenn2wang** [comment]
This is an interesting feature and very useful.
In my hypothesis, I want to share to specific users, and the user list is selectable by the current user.

**Comment by deansheather** [comment]
This would be a pretty big undertaking so I don't think we should do this unless it's actually important. Most of our signed app token issuance code would have to be rewritten and significant changes would have to be made to the RBAC engine.

**Comment by bpmct** [comment]
Yeah, I'd rather we start with making the coder_app sharing configurable per-workspace, not per-template #11486 (which we have prioritized for this quarter)

**Comment by matifali** [comment]
To summarize, the ideal scenario will be to expose this feature on the workspace page with settings if this can be done controlled per template. 
No deployment wide flags.

---

On a side note, we slowly move most and then all of the template settings and metadata all `coder push|edit` arguments to the Template terraform code. That would allow full control of templates as IaC. 

cc: @ammario and @bpmct for comments

**Comment by kylecarbs** [comment]
We're already doing this with shared ports!

**Comment by matifali** [comment]
duplicate of #13005

**Comment by bjornrobertsson** [comment]
Port sharing should also be granted on Group, Organisation and possibly a Role made available with wider App/Port visibility.
The Port sharing currently made with the 'Authenticated' method bypasses the need to keep a Workspace within one Org/Group.

**Comment by bpmct** [comment]
No plans to do this at this time, but we will be doing this with sharing of the overall workspace #2425
